arch/x86_64/inject.c

```diff
@@ -0,0 +1,205 @@
+/*
+ * Linux module for injecting shared object to the process by using ptrace.
+ *
+ * copied from: https://github.com/gaffe23/linux-inject
+ * Released under the GPL v2+.
+ */
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/user.h>
+#include <wait.h>
+#include <sys/time.h>
+#include "utils/inject.h"
+#include "utils/ptrace.h"
+#include "utils/utils.h"
+
+#define SO_LIBC_NAME "libc.so.6"
+#define SO_LIBDL_NAME "libdl.so.2"
+
+extern void inject_content(void);
+extern const void inject_content_end;
+
+int inject(char *libname, pid_t pid)
+{
+	struct user_regs_struct oldregs, regs;
+	struct user_regs_struct malloc_regs;
+	struct user_regs_struct dlopen_regs;
+	int lib_path_length, mypid;
+	long my_libc_addr, my_libdl_addr;
+	long malloc_addr, free_addr, dlopen_addr;
+	long malloc_offset, free_offset, dlopen_offset;
+	long target_libc_addr, target_libdl_addr; 
+	long target_malloc_addr, target_free_addr;
+	long target_dlopen_addr, addr;
+	unsigned long long target_buf, lib_addr;
+	char *newcode, *backup;
+	char *lib_path = realpath(libname, NULL);
+	pid_t target = 0;
+
+	// figure out the size of contents to be injected.
+	// so, we know how big of a buffer to allocate.
+	uint32_t inject_content_size = (uintptr_t)&inject_content_end - (uintptr_t)inject_content;
+	size_t inject_shared_lib_size = inject_content_size;
```

Comment:
No need to keep 2 variables for a same value IMO.  Also how about using a shorter name like "inject_size"?